File: solacher/src/jobs/tvl.js
require('dotenv').config()

const { parentPort: workerParent, workerData } = require('node:worker_threads')

const redis = require('@redis/client')
const request = require('got-cjs').got

const { pickAsWith } = require('../lib/utils')

async function fetchData() {
  return Promise.all([
    request('https://api.llama.fi/charts/Solana').json(),
    request('https://api.llama.fi/protocols').json(),
  ])
}

function normalizeData([rawHistory, rawProtocols]) {
  const history = rawHistory.map(item => pickAsWith([
    ['date', 'ts', x => parseInt(x) * 1000],
    ['totalLiquidityUSD', 'tvl'],
  ], item))
  const totalTvl = rawHistory[rawHistory.length-1].totalLiquidityUSD

  const protocols = rawProtocols
    .filter(protocol => protocol.chains.includes('Solana') && protocol.chainTvls && protocol.chainTvls['Solana'])
    .map(protocol => ({
      ...pickAsWith([
        ['category', 'category', x => x === 'Dexes' ? 'DEX' : x],
        'change_1h',
        'change_7d',
        ['change_1d', 'change_24h'],
        'description',
        ['listedAt', 'listedAt', x => parseInt(x) * 1000],
        ['logo', 'imageUrl'],
        ['mcap', 'marketCap'],
        'name',
        'symbol',
        'twitter',
        'url',
      ], protocol),
      dominancePercent: (protocol.chainTvls['Solana'] / totalTvl) * 100,
      tvl: protocol.chainTvls['Solana'],
    }))
    .sort((a, b) => (b.tvl || -1) - (a.tvl || -1))

  return {
    data: {
      history,
      totalTvl,
      protocols,
      protocolsCount: protocols.length,
    },
    count: null,
    type: 'object',
    updatedAt: Date.now(),
  }
}

;(async function main() {
  const redisUrl = workerParent
    ? workerData.data.redisUrl
    : process.env.REDIS_URL

  let data = null
  try {
    data = await fetchData()
      .then(normalizeData)
      .then(JSON.stringify)
  } catch (error) {
    console.error('Request(s) / processing failed for job "tvl"', error)
    if (workerParent) workerParent.postMessage('error')
    else process.exit(1)
  }

  const redisClient = redis.createClient({ url: redisUrl })
  await redisClient.connect()
  await redisClient.set('tvl', data)
  await redisClient.quit()

  if (workerParent) workerParent.postMessage('done')
  else process.exit(0)
})();

File: etl/etl/jobs/tvl.py
import json

import httpx
from dotenv import load_dotenv
from toolz import compose_left
from toolz.curried import filter, map, merge

from etl.connector import get_redis_connection
from etl.utils import now, pick, s_to_ms, sort

def normalize_protocol(raw_protocol, total_tvl, n):
    return pick(
        [
            ["symbol", "symbol", lambda s, _: None if s == "-" else s],
            "name",
            "description",
            ["logo", "imageUrl"],
            ["twitter", "urlTwitter"],
            ["url", "urlWebsite"],
            ["listedAt", "listedAt", lambda val, _: s_to_ms(val)],
            ["mcap", "marketCap"],
            ["name", "rank", lambda val, dic: n + 1],
            [
                "category",
                "category",
                lambda c, _: "DEX" if c == "Dexes" else c,
            ],
            [
                "chainTvls",
                "tvl",
                lambda _, dic: {
                    "current": dic["chainTvls"]["Solana"],
                    "change_7day": dic["change_7d"],
                    "change_24h": dic["change_1d"],
                    "change_1h": dic["change_1h"],
                    "dominancePercent": (dic["chainTvls"]["Solana"] / total_tvl)
                    * 100,
                },
            ],
        ],
        raw_protocol,
        enforce_unset=True,
    )

def normalize_tvl(raw_history, raw_protocols):
    total_tvl = raw_history[-1]["totalLiquidityUSD"]
    return {
        "totalTvl": total_tvl,
        "history": list(
            map(
                pick(
                    [
                        ["date", "ts", lambda val, _: s_to_ms(val)],
                        ["totalLiquidityUSD", "tvl"],
                    ]
                ),
                raw_history,
            )
        ),
        "protocols": list(
            compose_left(
                filter(
                    lambda raw_protocol: "Solana" in raw_protocol["chains"]
                    and "chainTvls" in raw_protocol
                    and "Solana" in raw_protocol["chainTvls"]
                ),
                sort(lambda a, b: b["chainTvls"]["Solana"] - a["chainTvls"]["Solana"]),
                lambda raw_protocols: [
                    normalize_protocol(raw_protocol, total_tvl, n) for n, raw_protocol in enumerate(raw_protocols)
                ]
            )(raw_protocols)
        ),
    }


def update_tvl():
    redis = get_redis_connection()


    tvl_response = httpx.get("https://api.llama.fi/charts/Solana")
    tvl_response.raise_for_status()
    raw_tvl_history = tvl_response.json()

    protocols_response = httpx.get("https://api.llama.fi/protocols")
    protocols_response.raise_for_status()
    raw_protocols = protocols_response.json()

    result = json.dumps(
        {
            "data": normalize_tvl(raw_tvl_history, raw_protocols),
            "count": None,
            "type": "object",
            "updatedAt": now(),
        }
    )

    redis.set("tvl", result)
    redis.close()

    return result


if __name__ == "__main__":
    load_dotenv()
    print(update_tvl())
